Add model doc for ViTPose with quantization and attention visualization

saumanraaj opened the pull request:
# What does this PR do?
This PR adds a new model documentation file for ViTPose, following the standardized doc format seen in models like BERT
<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->




## Before submitting
- [✓ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ✓] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [✓ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ✓] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).



## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker
- vision models: @amyeroberts, @qubvel
- speech models: @eustlb
- graph models: @clefourrier

Library:

- flax: @gante and @Rocketknight1
- generate: @zucchini-nlp (visual-language models) or @gante (all others)
- pipelines: @Rocketknight1
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @zach-huggingface and @SunMarc
- chat templates: @Rocketknight1

Integrations:

- deepspeed: HF Trainer/Accelerate: @SunMarc @zach-huggingface
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc @MekkCyber

Documentation: @stevhliu

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @Rocketknight1
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

saumanraaj committed:
Add model doc for ViTPose with quantization and attention visualization

saumanraaj committed:
Merge branch 'main' into add-vitpose-doc

saumanraaj commented:
Hi @stevhliu @qubvel! 👋

This PR adds standardized documentation for the ViTPose model, including quantization and attention visualization examples, consistent with other model cards like BERT.

Would you mind reviewing when you get a chance? Thank you so much!

saumanraaj commented:
The CI failed due to a kenlm build error. My PR only updates documentation for ViTPose, so this error isn’t caused by my changes.

saumanraaj committed:
Merge branch 'main' into add-vitpose-doc

qubvel commented on docs/source/en/model_doc/vitpose.md:
````diff
@@ -10,277 +10,107 @@ an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express o
-    outputs = person_model(**inputs)
+```python
+from transformers import pipeline
 
-results = person_image_processor.post_process_object_detection(
-    outputs, target_sizes=torch.tensor([(image.height, image.width)]), threshold=0.3
+pose_estimator = pipeline(
+    task="pose-estimation",
+    model="usyd-community/vitpose-base-simple"
 )
````
We dont have such a pipeline

qubvel reviewed (changes requested):
Hey @saumanraaj, thanks for the contribution, we should not remove relevant information from the VitPose model doc, in addition, we should keep **working** code snippets

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_37089). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

stevhliu commented on docs/source/en/model_doc/vitpose.md:
````diff
@@ -10,277 +10,107 @@ an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express o
 
-```python
-import supervision as sv
+---
````
```suggestion
```

stevhliu replied:
Need to remove this too

stevhliu commented on docs/source/en/model_doc/vitpose.md:
```diff
@@ -10,277 +10,107 @@ an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express o
 
-xy = torch.stack([pose_result['keypoints'] for pose_result in image_pose_result]).cpu().numpy()
-scores = torch.stack([pose_result['scores'] for pose_result in image_pose_result]).cpu().numpy()
+## Quantization Example
```
Probably better to use a different quantization backend, like torchao, that isn't specific to Intel hardware. Remember to add the complete code snippet for using quantization!

```suggestion
Quantization reduces the memory burden of large models by representing the weights in a lower precision. Refer to the [Quantization](https://huggingface.co/docs/transformers/main/en/quantization/overview) overview for more available quantization backends.

The example below uses [torchao](https://huggingface.co/docs/transformers/main/en/quantization/torchao) to only quantize the weights to int4.
```

stevhliu replied:
Remove `## Quantization Example` and add the text above!

stevhliu commented on docs/source/en/model_doc/vitpose.md:
```diff
@@ -10,277 +10,107 @@ an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express o
-for pose_result in image_pose_result:
-    scores = np.array(pose_result["scores"])
-    keypoints = np.array(pose_result["keypoints"])
+---
```
```suggestion
```

stevhliu replied:
Remove!

stevhliu commented on docs/source/en/model_doc/vitpose.md:
```diff
@@ -10,277 +10,107 @@ an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express o
 
-    # draw each point on image
-    draw_points(numpy_image, keypoints, scores, keypoint_colors, keypoint_score_threshold=0.3, radius=4, show_keypoint_weight=False)
+## Attention Visualization
```
The attention mask visualizer isn't supported for ViTPose yet

stevhliu commented on docs/source/en/model_doc/vitpose.md:
```diff
@@ -10,277 +10,107 @@ an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express o
-<img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/transformers/model_doc/vitpose-coco.jpg" alt="drawing" width="600"/>
 
-## Resources
+---
```
```suggestion
```

stevhliu replied:
Remove!

stevhliu commented on docs/source/en/model_doc/vitpose.md:
```diff
@@ -10,277 +10,107 @@ an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express o
 
-- A demo of ViTPose on images and video can be found [here](https://huggingface.co/spaces/hysts/ViTPose-transformers).
-- A notebook illustrating inference and visualization can be found [here](https://github.com/NielsRogge/Transformers-Tutorials/blob/master/ViTPose/Inference_with_ViTPose_for_human_pose_estimation.ipynb).
+---
```
```suggestion
```

stevhliu replied:
Remove!

stevhliu commented on docs/source/en/model_doc/vitpose.md:
```diff
@@ -10,277 +10,107 @@ an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express o
+## Notes
 
-A list of official Hugging Face and community (indicated by 🌎) resources to help you get started with ViTPose. If you're interested in submitting a resource to be included here, please feel free to open a Pull Request and we'll review it! The resource should ideally demonstrate something new instead of duplicating an existing resource.
+- ViTPose is a top-down pose estimator: it requires a prior object detector to crop individuals before keypoint prediction.
```
```suggestion
- ViTPose is a top-down pose estimator. It uses a object detector to detect individuals first before keypoint prediction.
```

stevhliu commented on docs/source/en/model_doc/vitpose.md:
```diff
@@ -10,277 +10,107 @@ an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express o
 
-A list of official Hugging Face and community (indicated by 🌎) resources to help you get started with ViTPose. If you're interested in submitting a resource to be included here, please feel free to open a Pull Request and we'll review it! The resource should ideally demonstrate something new instead of duplicating an existing resource.
+- ViTPose is a top-down pose estimator: it requires a prior object detector to crop individuals before keypoint prediction.
+- ViTPose++ supports passing a `dataset_index` to activate different MoE expert heads.
```
Add links to the datasets and include an indented code snippet demonstrating how to pass the `dataset_index`

```suggestion
- ViTPose++ has 6 different MoE expert heads (COCO validation, AiC, MPII, AP-10K, APT-36K, COCO-WholeBody) which supports 6 different datasets. You can pass a specific dataset index to the model to indicate which expert to use.
```

stevhliu commented on docs/source/en/model_doc/vitpose.md:
```diff
@@ -10,277 +10,107 @@ an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express o
-A list of official Hugging Face and community (indicated by 🌎) resources to help you get started with ViTPose. If you're interested in submitting a resource to be included here, please feel free to open a Pull Request and we'll review it! The resource should ideally demonstrate something new instead of duplicating an existing resource.
+- ViTPose is a top-down pose estimator: it requires a prior object detector to crop individuals before keypoint prediction.
+- ViTPose++ supports passing a `dataset_index` to activate different MoE expert heads.
+- Use `AutoProcessor` to handle bounding box preprocessing automatically.
+- Ideal for use cases involving MS COCO, WholeBody, MPII, AP-10K, and APT-36K keypoint datasets.
```
Use the `<hfoptions>...</hfoptions>` tags under this to add code snippets for supervision and OpenCV.

```suggestion
- Use the [supervision](https://github.com/roboflow/supervision) or [OpenCV](https://opencv.org/) libraries to visualize the keypoints.
```

stevhliu reviewed (commented):
Thanks for starting this! Left some comments to include relevant code examples and picking a different quantization backend.

saumanraaj committed:
Merge branch 'main' into add-vitpose-doc

saumanraaj commented:
Final Updates Applied – Ready for Review
Hi @stevhliu, @qubvel  thank you for your thoughtful and detailed feedback. I’ve reviewed and addressed all the requested changes. Below is a summary of the updates made:

Cleanup of Deprecated or Invalid Examples
Removed the unsupported pipeline(...) usage for pose estimation.

Removed the transformers-cli example, which does not support image inputs.

Removed the AttentionMaskVisualizer block, as it is not compatible with ViTPose.

Removed the outdated usage snippet involving AutoProcessor and VitPoseForPoseEstimation without bounding box preprocessing.

Updated Code and Documentation
Rewrote the main usage section with a complete two-stage pipeline using RT-DETR for person detection and ViTPose for keypoint estimation.

Added support for dataset_index in ViTPose++ models, with linked datasets and an example showing how to pass the appropriate index.

Updated all model loading to use AutoModelForObjectDetection and AutoModelForPoseEstimation for consistency with Transformers conventions.

Replaced the Intel-specific quantization backend (optimum.intel) with torchao, and included a clear example demonstrating int4 quantization.

Added keypoint visualization examples using both the supervision library and OpenCV with customizable color palettes and connections.

Revised the Resources section to include links to the Hugging Face Spaces demo, a detailed notebook, and the original GitHub codebase.

Clarified and expanded the Notes section to highlight the top-down architecture, preprocessing with AutoProcessor, and relevant datasets.

qubvel commented:
Nothing have been changed btw

saumanraaj committed:
✅ Final update: full ViTPose doc with all fixes

saumanraaj committed:
Merge branch 'main' into add-vitpose-doc

saumanraaj commented:
Hi @stevhliu  @qubvel ,

I've applied all the suggested changes and finalized the ViTPose model card. The updates include:

Cleaned up outdated usage examples and unsupported pipelines

Added complete working usage with AutoProcessor and AutoModelForPoseEstimation

Included a quantization section using torchao

Added keypoint visualization examples using both supervision and OpenCV

Clarified dataset_index usage for ViTPose++ with linked datasets

Updated notes, citations, and formatting for clarity and consistency

Please let me know if there's anything else I should address. Thanks again for the detailed feedback!

Best,
Sauman

stevhliu commented on docs/source/en/model_doc/vitpose.md:
```diff
@@ -10,277 +10,107 @@ an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express o
-# ------------------------------------------------------------------------
-# Stage 1. Detect humans on the image
-# ------------------------------------------------------------------------
+---
```
This hasn't been resolved yet!

stevhliu commented on docs/source/en/model_doc/vitpose.md:
```diff
@@ -1,286 +1,151 @@
-inputs = image_processor(image, boxes=[person_boxes], return_tensors="pt").to(device)
-
-dataset_index = torch.tensor([0], device=device) # must be a tensor of shape (batch_size,)
+import torch
```
You can use the `TorchAoConfig` here instead, see the [Gemma 3](https://huggingface.co/docs/transformers/main/en/model_doc/gemma3#gemma-3) for an example

stevhliu commented on docs/source/en/model_doc/vitpose.md:
```diff
@@ -289,4 +154,20 @@ A list of official Hugging Face and community (indicated by 🌎) resources to h
\ No newline at end of file
+  - forward
+
+---
```
Remove and the citation under it!

stevhliu commented on docs/source/en/model_doc/vitpose.md:
```diff
@@ -1,286 +1,151 @@
-- 5: [COCO-WholeBody](https://github.com/jin-s13/COCO-WholeBody) dataset
+---
 
+## Keypoint Visualization
```
This should go under the `## Notes` section and it doesn't need its own separate section level

stevhliu commented on docs/source/en/model_doc/vitpose.md:
```diff
@@ -1,286 +1,151 @@
-    # draw links
-    draw_links(numpy_image, keypoints, scores, keypoint_edges, link_colors, keypoint_score_threshold=0.3, thickness=1, show_keypoint_weight=False)
+- ViTPose is a top-down pose estimator, meaning it requires a separate object detector to first locate individuals in the image.
+- Use `AutoProcessor` to automatically prepare bounding box and image inputs.
```
```suggestion
- Use [`AutoProcessor`] to automatically prepare bounding box and image inputs.
```

stevhliu commented on docs/source/en/model_doc/vitpose.md:
```diff
@@ -1,286 +1,151 @@
-    draw_links(numpy_image, keypoints, scores, keypoint_edges, link_colors, keypoint_score_threshold=0.3, thickness=1, show_keypoint_weight=False)
+- ViTPose is a top-down pose estimator, meaning it requires a separate object detector to first locate individuals in the image.
+- Use `AutoProcessor` to automatically prepare bounding box and image inputs.
+- ViTPose++ supports passing a `dataset_index` parameter to specify which expert head to use.
```
Recommend clarifying the `dataset_index` parameter and how they relate to the expert heads. See my comment [here](https://github.com/huggingface/transformers/pull/37089#discussion_r2021544579)

stevhliu replied:
Add a code snippet here demonstrating using `dataset_index` with ViTPose++ models to fix #36773!

stevhliu commented on docs/source/en/model_doc/vitpose.md:
```diff
@@ -1,286 +1,151 @@
-
-To visualize the various keypoints, one can either leverage the `supervision` [library](https://github.com/roboflow/supervision (requires `pip install supervision`):
+<hfoptions id="visualization">
+<hfoption id="Supervision">
```
```suggestion
<hfoption id="supervision">
```

stevhliu commented on docs/source/en/model_doc/vitpose.md:
```diff
@@ -1,286 +1,151 @@
 [[autodoc]] VitPoseImageProcessor
-    - preprocess
-    - post_process_pose_estimation
+  - preprocess
```
No need to change the indentation level of these methods
